Sources/System/MachPort.swift

```diff
@@ -0,0 +1,318 @@
+                    // recv rights must be mod ref'ed instead of deallocated
+                    let kr = mach_port_unguard(mach_task_self_, name, context)
+                    assert(kr == KERN_SUCCESS)
+
+                    let kr2 = mach_port_mod_refs(mach_task_self_, name, MACH_PORT_RIGHT_RECEIVE, -1)
+                    assert(kr2 == KERN_SUCCESS)
+                } else {
+                    mach_port_deallocate(mach_task_self_, name)
+                }
+            }
+        }
+    }
+
+    /// Possible errors that can be thrown by Mach.Port operations.
+    enum PortRightError : Error {
+        /// Returned when an operation cannot be completed, because the Mach
+        /// port right has become a dead name. This is caused by deallocation of the
+        /// receive right on the other end.
+        case deadName
+    }
+
+    /// The MachPortRight type used to manage a receive right.
+    struct ReceiveRight : MachPortRight {}
+
+    /// The MachPortRight type used to manage a send right.
+    struct SendRight : MachPortRight {}
+
+    /// The MachPortRight type used to manage a send-once right.
+    ///
+    /// Send-once rights are the most restrictive type of Mach port rights.
+    /// They cannot create other rights, and are consumed upon use.
+    ///
+    /// Upon destruction a send-once notification will be sent to the
+    /// receiving end.
+    struct SendOnceRight : MachPortRight {}
+
+    /// Create a connected pair of rights, one receive, and one send.
+    ///
+    /// This function will abort if the rights could not be created.
+    /// Callers may assert that valid rights are always returned.
+    static func allocatePortRightPair() -> (Mach.Port<Mach.ReceiveRight>, Mach.Port<Mach.SendRight>) {
+        var name = mach_port_name_t(MACH_PORT_NULL)
+        let secret = mach_port_context_t(arc4random())
+        withUnsafeMutablePointer(to: &name) { name in
+            var options = mach_port_options_t()
+            options.flags = UInt32(MPO_INSERT_SEND_RIGHT);
+            withUnsafeMutablePointer(to: &options) { options in
+                let kr = mach_port_construct(mach_task_self_, options, secret, name)
+                assert(kr == KERN_SUCCESS)
+            }
+        }
+        return (Mach.Port<Mach.ReceiveRight>(name: name, context: secret), Mach.Port<Mach.SendRight>(name: name))
+    }
+}
+
+extension Mach.Port where RightType == Mach.ReceiveRight {
+    /// Transfer ownership of an existing, unmanaged, but already guarded,
+    /// Mach port right into a Mach.Port by name.
+    ///
+    /// This initializer aborts if name is MACH_PORT_NULL.
+    ///
+    /// If the type of the right does not match the type T of Mach.Port<T>
+    /// being constructed, the behavior is undefined.
+    ///
+    /// The underlying port right will be automatically deallocated when
+    /// the Mach.Port object is destroyed.
+    init(name:mach_port_name_t, context:mach_port_context_t) {
```

Comment:
```suggestion
    init(name: mach_port_name_t, `guard`: mach_port_context_t) {
```

Reply:
I fixed the whitespace, but kept this as `context` both here and in the new tuple label.